src/Compilers/CSharp/Portable/Symbols/Metadata/PE/NativeIntegerTypeDecoder.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Reflection.Metadata;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
+{
+    internal struct NativeIntegerTypeDecoder
+    {
+        internal static TypeSymbol TransformType(TypeSymbol type, EntityHandle handle, PEModuleSymbol containingModule)
+        {
+            return containingModule.Module.HasNativeIntegerAttribute(handle, out var transformFlags) ?
+                TransformType(type, transformFlags) :
+                type;
+        }
+
+        private static TypeSymbol TransformType(TypeSymbol type, ImmutableArray<bool> transformFlags)
+        {
+            var decoder = new NativeIntegerTypeDecoder(transformFlags);
+            try
+            {
+                var result = decoder.TransformType(type);
+                if (decoder._index == transformFlags.Length)
```

Comment:
If this is false does that mean the metadata is malformed?

Reply:
Yes, this would fail if there are more bools in the array than we expect. And the case where are fewer bools is handled in `Increment()`.

---
In reply to: [373660152](https://github.com/dotnet/roslyn/pull/41160#discussion_r373660152) [](ancestors = 373660152)